### PR TITLE
Add option to use custom dataLayer variable

### PR DESCRIPTION
**Pull request by eljass**
This pull request introduces the ability to change the name of the `dataLayer` used by GTM. The new feature allows passing an option called `dataLayerName`, which initializes a window variable that contains the dataLayer array and changes helper methods such as `push()`, `trackEvent()`, and `trackView()` to push data to this renamed dataLayer variable.

When fetching Tag Manager's script, this uses Google's recommendation to pass the dataLayerName as a query parameter.

### CODE REVIEW

File: package-lock.json [hunk 1]
```diff
@@ -0,0 +1,6406 @@
+{
```
Comment (Shinigami92) [inline package-lock.json:1]: This project uses pnpm, please delete this file because pnpm-lock.yaml is already used.

You might need to use `corepack enable` on your machine

File: package.json [hunk 2]
```diff
@@ -8,7 +8,7 @@
     "format": "prettier --cache --write .",
     "lint": "eslint --cache --cache-strategy content --report-unused-disable-directives .",
     "ts-check": "tsc",
-    "test": "vitest",
+    "test": "vitest --silent=false",
```
Comment (Shinigami92) [inline package.json:11]: This change is not related to the feature request
please revert it

File: package.json [hunk 3]
```diff
@@ -62,7 +62,7 @@
     "rimraf": "~5.0.1",
     "tsup": "~7.2.0",
     "typescript": "~5.2.2",
-    "vitest": "~0.34.5"
+    "vitest": "^0.34.6"
```
Comment (Shinigami92) [inline package.json:65]: This change is not related to the feature request
please revert it

File: package.json [hunk 4]
```diff
@@ -62,7 +62,7 @@
...
   },
   "packageManager": "pnpm@8.7.6"
-}
+}
```
Comment (Shinigami92) [inline package.json:68]: please run `pnpm run format` so the project is formatted correctly

File: src/data-layer-object.ts [hunk 5]
```diff
@@ -1,5 +1,5 @@
 /**
- * Object within the `window.dataLayer`.
+ * Object within the `window[pluginOptions[dataLayerName]]`.
```
Comment (Shinigami92) [inline src/data-layer-object.ts:2]: I think this change is not really relevant.
We should only document necessary parts with that custom change.

File: src/gtm-support.ts [hunk 6]
```diff
@@ -144,14 +145,18 @@ export class GtmSupport {
   }
 
   /**
-   * Returns the `window.dataLayer` array if the script is running in browser context and the plugin is enabled,
+   * Returns the `window[pluginOptions[dataLayerName]]` array if the script is running in browser context and the plugin is enabled,
```
Comment (Shinigami92) [inline src/gtm-support.ts:148]: This is not really correct, I think it is more like `window[dataLayerName]`

File: src/gtm-support.ts [hunk 7]
```diff
@@ -144,14 +145,18 @@ export class GtmSupport {
...
    * otherwise `false`.
    *
-   * @returns The `window.dataLayer` if script is running in browser context and plugin is enabled, otherwise `false`.
+   * @returns The `window[pluginOptions[dataLayerName]]` if script is running in browser context and plugin is enabled, otherwise `false`.
```
Comment (Shinigami92) [inline src/gtm-support.ts:151]: same here

File: src/index.ts [hunk 8]
```diff
@@ -3,7 +3,14 @@ import type { DataLayerObject } from './data-layer-object';
 declare global {
   interface Window {
     /**
-     * `dataLayer` used by GTM.
+     * Dynamic `dataLayer` key used by GTM.
+     *
+     * @see [developers.google.com/tag-manager/devguide/datalayer#rename_the_data_layer](https://developers.google.com/tag-platform/tag-manager/datalayer#rename_the_data_layer)
+     */
+    [key: string]: any;
```
Comment (Shinigami92) [inline src/index.ts:10]: This is a really bad idea. Instead we should document how to add TS support for a custom key if really necessary/wanted.

File: src/options.ts [hunk 9]
```diff
@@ -90,4 +90,12 @@ export interface GtmSupportOptions {
    * @default content-view
    */
   trackViewEventProperty?: string;
+  /**
+   * The name of the dataLayer variable to push events.
+   *
+   * @example dataLayerName: 'dataLayerDemo'
+   *
+   * @default dataLayer
+   */
```
Comment (Shinigami92) [inline src/options.ts:99]: we could add an `@see https://developers.google.com/tag-platform/tag-manager/datalayer#rename_the_data_layer` to this JSDoc :thinking:

File: src/options.ts [hunk 10]
```diff
@@ -90,4 +90,12 @@ export interface GtmSupportOptions {
    * @default content-view
    */
   trackViewEventProperty?: string;
+  /**
+   * The name of the dataLayer variable to push events.
+   *
+   * @example dataLayerName: 'dataLayerDemo'
+   *
+   * @default dataLayer
```
Comment (Shinigami92) [inline src/options.ts:98]:
```suggestion
   * @default 'dataLayer'
```

File: src/utils.ts [hunk 11]
```diff
@@ -58,6 +58,12 @@ export interface LoadScriptOptions {
    * @param options Object containing container `id` and `script` element.
    */
   onReady?: (options: OnReadyOptions) => void;
+  /**
+   * The GTM dataLayer name.
+   *
+   * @default 'dataLayer'
+   */
```
Comment (Shinigami92) [inline src/utils.ts:65]: We can add that `@see` here as well.

File: src/utils.ts [hunk 12]
```diff
@@ -82,9 +88,10 @@ export function loadScript(
 
   script.addEventListener('load', scriptLoadListener);
 
-  window.dataLayer = window.dataLayer ?? [];
+  const dataLayerName: string = config.dataLayerName ?? 'dataLayer';
+  window[dataLayerName] = window[dataLayerName] ?? [];
 
-  window.dataLayer?.push({
+  window[dataLayerName]?.push({
```
Comment (Shinigami92) [inline src/utils.ts:94]: mh :thinking: not really related to your MR, but I think we could safely use

```suggestion
  window[dataLayerName]!.push({
```

here

File: src/utils.ts [hunk 13]
```diff
@@ -106,6 +113,7 @@ export function loadScript(
 
   const queryString: URLSearchParams = new URLSearchParams({
     id,
+    l: config.dataLayerName ?? 'dataLayer',
```
Comment (Shinigami92) [inline src/utils.ts:116]: IMO we should add the `l` query param only when it is **not** equals `'dataLayer'`

**Review by Shinigami92 — CHANGES_REQUESTED** [review]

**Comment by eljass** [comment]
@Shinigami92 I refactored the code based on your reviews.

Those made sense, and I'm sorry for not noticing the use of `pnpm` and the fixes to take such a long time. The main issue I experienced with this refactor was the change in the global window declaration. We'll somehow need to handle the types for the window so we can use the dynamic key for the `dataLayer` name. The solution is to have an extended interface for the window, which is used only inside this module and not passed on to the end product. I am not sure if documentation for typing the window is in the scope of this module or change as if the end user is using typescript, they should handle the window's types by themselves.

Let me know if the changes make sense for you, as I did create a reusable function to get or construct the `window.dataLayer` array. Otherwise, there would be lot of repetition in defining window type to allow dynamic naiming of the `dataLayer`. I did also refactored tests to use `dataLayer` from `instance`, to avoid same repetition. I don't know if you had some other justifications to not do so previously?

File: src/utils.ts (hunk 12 above)
Comment (eljass) [reply src/utils.ts]: Sorted this out by getting or initializing dataLayer in reusable function

**Review by Shinigami92 — APPROVED** [review]
